### PR TITLE
tooling: include OSV JSON data in official CVE feed

### LINKED ISSUE #177: tooling: add the OSV CVE information to the official CVE feed
You can find the full context here https://github.com/kubernetes/sig-security/issues/169.

We have the official CVE feed which is pretty well documented:
- https://github.com/kubernetes/sig-security/issues/1
- https://kubernetes.io/docs/reference/issues-security/official-cve-feed/
- https://github.com/kubernetes/sig-security/tree/main/sig-security-tooling/cve-feed

The idea would be to add the OSV information from:
- [ ] https://github.com/kubernetes/sig-security/issues/179 
  **The new OSV data for new CVEs from the issues**, basically we merged this new tool used by the SRC members to publish the CVE announcement so that it contains a `json osv` JSON blog. See more information to parse the information in the PR: https://github.com/kubernetes/sig-security/pull/171. You can see how the output looks like, it's been recently used on:
  - https://github.com/kubernetes/kubernetes/issues/136680
  - https://github.com/kubernetes/kubernetes/issues/136677
  - https://github.com/kubernetes/kubernetes/issues/136678
  - https://github.com/kubernetes/kubernetes/issues/136679
  - https://github.com/kubernetes/kubernetes/issues/136789
- [ ] https://github.com/kubernetes/sig-security/issues/178
  **The legacy OSV data for the previous CVEs from https://github.com/kubernetes-sigs/cve-feed-osv**, we have this existing database of CVE OSV information that we could enrich the JSON feed with https://github.com/kubernetes-sigs/cve-feed-osv/tree/main/vulns.

## How to merge the OSV scheme into the JSON feed

Here's my idea, maybe you'll have a better one!

I don't know if there's an official version of a feed of OSV format CVEs, but I think the least invasive change we could do is to add a new custom field in existing the JSON feed. Custom fields are starting with `_`, like the `_kubernetes_io` (based on the JSON feed spec https://www.jsonfeed.org/version/1/) you can see in items in https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json. 

```json5
{
  "_kubernetes_io": {
    "feed_refresh_job": "https://testgrid.k8s.io/sig-security-cve-feed#auto-refreshing-official-cve-feed",
    "updated_at": "2026-02-06T14:16:51Z"
  },
  "authors": [
    {
      "name": "Kubernetes Community",
      "url": "https://www.kubernetes.dev"
    }
  ],
  "description": "Auto-refreshing official CVE feed for Kubernetes repository",
  "feed_url": "https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json",
  "home_page_url": "https://kubernetes.io",
  "items": [
    {
      "_kubernetes_io": {
        "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2025-15566",
        "issue_number": 136789
      },
      "content_text": "CVSS Rating: [CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H](https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H)\n\nA security issue was discovered in ingress-nginx [...] by Jan-Otto Kröpke.\",\n      \"type\": \"FINDER\"\n    }\n  ]\n}\n```\n\n\u003C/details\u003E\n\n\u003C!-- generated by srctl v1.0.0 (2665728667a5-dirty, 2026-01-31T19:18:07Z, go1.25.6) --\u003E\n",
      "date_published": "2026-02-06T02:54:24Z",
      "external_url": "https://www.cve.org/cverecord?id=CVE-2025-15566",
      "id": "CVE-2025-15566",
      "status": "fixed",
      "summary": "ingress-nginx auth-proxy-set-headers nginx configuration injection",
      "url": "https://github.com/kubernetes/kubernetes/issues/136789"
    },
    // [...]
}
```

So for example, what could be done is to ship the OSV thing under `_kubernetes_io` like:
```json5
{
    "_kubernetes_io": {
      "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2025-15566",
      "issue_number": 136789
      "osv": { // <--- full OSV object inserted here
        "schema_version": "1.6.0",
        "id": "CVE-2025-15566",
        "modified": "2026-02-06T02:46:50Z",
        "summary": "ingress-nginx auth-proxy-set-headers nginx configuration injection",
        "details": "A security issue was discovered in ingress-nginx where the nginx.ingress.kubernetes.io/auth-proxy-set-headers Ingress annotation can be used to inject configuration into nginx. This can lead to arbitrary code execution in the context of the ingress-nginx controller, and disclosure of Secrets accessible to the controller. (Note that in the default installation, the controller can access all Secrets cluster-wide.)",
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H"
          }
        ],
        "affected": [
          {
            "package": {
              "ecosystem": "Kubernetes",
              "name": "ingress-nginx"
            },
            "ranges": [
              {
                "type": "SEMVER",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "v1.12.5"
                  },
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "v1.13.1"
                  }
                ]
              }
            ]
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H"
          }
        ],
        "credits": [
          {
            "name": "This issue was discoverd **and patched** by Jan-Otto Kröpke.",
            "type": "FINDER"
          }
        ]
      }
    },
    "content_text": "CVSS Rating: [CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H](https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H)\n\nA security issue was discovered in ingress-nginx [...] by Jan-Otto Kröpke.\",\n      \"type\": \"FINDER\"\n    }\n  ]\n}\n```\n\n\u003C/details\u003E\n\n\u003C!-- generated by srctl v1.0.0 (2665728667a5-dirty, 2026-01-31T19:18:07Z, go1.25.6) --\u003E\n",
    "date_published": "2026-02-06T02:54:24Z",
    "external_url": "https://www.cve.org/cverecord?id=CVE-2025-15566",
    "id": "CVE-2025-15566",
    "status": "fixed",
    "summary": "ingress-nginx auth-proxy-set-headers nginx configuration injection",
    "url": "https://github.com/kubernetes/kubernetes/issues/136789"
}
```

### LINKED ISSUE #178: tooling: include `kubernetes-sigs/cve-feed-osv` OSV information in the CVE feed
Sub-issue of https://github.com/kubernetes/sig-security/issues/177, see all the context and explanations there.

In short it consists of parsing the OSV content from https://github.com/kubernetes-sigs/cve-feed-osv/tree/main/vulns an including it in the JSON feed.

### LINKED ISSUE #179: tooling: include new CVE issues JSON OSV blob information in the CVE feed
Sub-issue of https://github.com/kubernetes/sig-security/issues/177, see all the context and explanations there.

In short it consists of parsing the new JSON OSV scheme included in the newest CVE announcement issues from the SRC in k/k and including it in the JSON feed.

### PR DESCRIPTION
**Pull request by ghost**
## What I Did

- Extended the official Kubernetes CVE feed to include OSV JSON data.
- For each CVE issue:
  - First attempt to extract an embedded OSV JSON object from the issue body
    (new SRC CVE announcement format).
  - Fallback to fetching the OSV JSON from the
    `kubernetes-sigs/cve-feed-osv` repository when no embedded OSV is present.
  - Added the OSV data under the `_kubernetes_io["osv"]` field.
  - Ensured that both the primary CVE and any additional CVEs referenced in the
    same issue include OSV data.
- Implemented safe network handling with `try/except` blocks and a `timeout=5`
  to prevent feed generation failures.
- Maintained backward compatibility: when no OSV data is available, the `osv`
  field remains `None`.

## Related Issue

- Fixes: #177
- Fixes: #178 
- Fixes: #179 

## Tests
I picked 2 different CVEs to do the test, in which the first one is related to the correct error, that is, a null return for an expected pattern.The second one is based on the correct pattern related to a CVE with the public OSV.

### CVE test with OSV not published
The correct and standard behavior is for the OSV to be null, since this example CVE does not have a published OSV.

image

### CVE with a Public OSV
CVE with public OSV where the expected default behavior is receiving a JSON with the correct information.

#### Output Terminal
```json
{
  "id": "CVE-2019-11243",
  "osv": {
    "id": "CVE-2019-11243",
    "modified": "2019-04-18T21:31:53Z",
    "published": "2019-04-18T21:31:53Z",
    "summary": "rest.AnonymousClientConfig() does not remove the serviceaccount credentials from config created by rest.InClusterConfig()",
    "details": "In Kubernetes v1.12.0-v1.12.4 and v1.13.0, the rest.AnonymousClientConfig() method returns a copy of the provided config, with credentials removed (bearer token, username/password, and client certificate/key data). In the affected versions, rest.AnonymousClientConfig() did not effectively clear service account credentials loaded using rest.InClusterConfig()",
    "affected": [
      {
        "package": {
          "ecosystem": "kubernetes",
          "name": "k8s.io/client-go"
        },
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:L/I:N/A:N"
          }
        ],
        "ranges": [
          {
            "type": "SEMVER",
            "events": [
              {
                "introduced": "1.12.0"
              },
              {
                "fixed": "1.12.5"
              },
              {
                "introduced": "1.13.0"
              },
              {
                "fixed": "1.13.1"
              }
            ]
          }
        ]
      }
    ],
    "references": [
      {
        "type": "ADVISORY",
        "url": "https://github.com/kubernetes/kubernetes/issues/76797"
      },
      {
        "type": "ADVISORY",
        "url": "https://www.cve.org/cverecord?id=CVE-2019-11243"
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by PushkarJ** [comment]
@Pnkcaht Thank you so much for making this contribution. I tested this locally and it works well!

/lgtm
/assign mtardy

**Comment by Pnkcaht** [comment]
@PushkarJ  Thank you my friend, I will be contributing regularly here as well.

File: sig-security-tooling/cve-feed/hack/__pycache__/cve_title_parser.cpython-313.pyc
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/__pycache__/cve_title_parser.cpython-313.pyc:1]: hey you committed a binary file here, please remove

File: sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py [hunk 1]
```diff
@@ -91,15 +95,63 @@ def getCVEStatus(state, state_reason):
 
         first_cve_id = cve_ids[0]
         cve['id'] = first_cve_id
+
+        # Initialize the OSV field as None in case no OSV JSON is available
+        cve['_kubernetes_io']['osv'] = None
```
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py:100]: you already did that on line 82

File: sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py [hunk 2]
```diff
@@ -91,15 +95,63 @@ def getCVEStatus(state, state_reason):
...
+            # Attempt to fetch the OSV JSON with a 5-second timeout
+            res_osv = requests.get(osv_url, timeout=5)
+
+            # If the OSV JSON exists, parse it and store it in the 'osv' field
+            if res_osv.status_code == 200:
+                cve['_kubernetes_io']['osv'] = res_osv.json()
+
+        except requests.RequestException:
+            # If any network error occurs (timeout, connection error, etc.), leave 'osv' as None
+            cve['_kubernetes_io']['osv'] = None
```
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py:115]: It's already none, you can again remove that

File: sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py [hunk 3]
```diff
@@ -91,15 +95,63 @@ def getCVEStatus(state, state_reason):
...
+            # Make a deep copy of the main CVE to avoid overwriting its data
             additional_cve = copy.deepcopy(cve)
+
+            # Update the CVE ID for this additional CVE
             additional_cve['id'] = additional_cve_id
+
+            # Set the external URL for this CVE on CVE.org
             additional_cve['external_url'] = f'https://www.cve.org/cverecord?id={additional_cve_id}'
+
+            # Set the Google Group URL specific to this CVE
```
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py:132]: please do not add unnecessary comments on un-related lines to your changes

File: sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py [hunk 4]
```diff
@@ -91,15 +95,63 @@ def getCVEStatus(state, state_reason):
...
+                # Attempt to fetch the OSV JSON with a 5-second timeout
+                res_additional_osv = requests.get(additional_osv_url, timeout=5)
+
+                # If the file exists, parse it as JSON and store it in the 'osv' field
+                if res_additional_osv.status_code == 200:
+                    additional_cve['_kubernetes_io']['osv'] = res_additional_osv.json()
+
+            except requests.RequestException:
+                # If any network error occurs (timeout, connection issue, etc.), keep 'osv' as None
+                additional_cve['_kubernetes_io']['osv'] = None
```
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py:152]: could you reuse that code with a function instead of duplicating?

File: sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py [hunk 5]
```diff
@@ -18,9 +18,29 @@
...
 
+def extract_osv_from_body(body):
+    # Extract an embedded OSV JSON object from a CVE issue body.
+    # New SRC CVE announcements may include the OSV data inline
+    # as a fenced ```json code block. This helper parses and returns
+    # that JSON when present.
+    if not body:
+        return None
+
+    match = re.search(r"```json\s*(\{.*?\})\s*```", body, re.DOTALL)
```
Comment (mtardy) [inline sig-security-tooling/cve-feed/hack/fetch-official-cve-feed.py:34]: please match the full "\`\`\`json osv" because the "\`\`\`json" can appear in different places in the issue if any json blob is included.

**Review by mtardy — CHANGES_REQUESTED** [review]
Thanks, it works but I think the patch could be adjusted

**Comment by mtardy** [comment]
> I picked 2 different CVEs to do the test, in which the first one is related to the correct error, that is, a null return for an expected pattern.The second one is based on the correct pattern related to a CVE with the public OSV.

Also by the way, could you add actual tests that indeed for one coming from the OSV feed repo you have a not null OSV value, and the same from the body of an issue.

**Comment by PushkarJ** [comment]
@mtardy many of the PR suggestions should be fixed in my follow up PR: https://github.com/kubernetes/sig-security/pull/183

I opened this one myself instead of getting those changes in as PR review feedback and delaying the PR merge.
